File: tracing/data-formats/src/data_formats/python/heracles/fhe_trace/io.py
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0


# TODO: create also C++ variants of below; given how simple and stable these functions should be just in replicated form, not shared code

from heracles.proto.fhe_trace_pb2 import Trace as TraceV1
from heracles.proto.fhe_trace_pb2 import Trace
import heracles.proto.common_pb2 as hpc
import heracles.proto.fhe_trace_pb2 as hpf
import csv

# load & store functions
# ===============================


def store_trace(filename: str, trace: Trace):
    """
    Serialize and store a HEC trace. Filename is constructed by concatenating `filename_prefix` with standard suffix.
    Prefix can contain directory paths, although they must all be existing directories
    """
    with open(filename, "wb") as f:
        f.write(trace.SerializeToString())


def load_trace(filename: str) -> Trace:
    """
    Load and deserialize a HEC trace. Filename is constructed by concatenating `filename_prefix` with standard suffix.
    Prefix can contain directory paths, although they must all be existing directories
    """
    trace = Trace()
    with open(filename, "rb") as f:
        trace.ParseFromString(f.read())
    return trace

File: tracing/data-formats/src/data_formats/cpp/heracles/fhe_trace/io.cpp
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "heracles/fhe_trace/io.h"
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>

namespace heracles::fhe_trace
{
bool store_trace(const std::string &filename, const heracles::fhe_trace::Trace &trace)
{
    std::ofstream pb_ofile(filename, std::ios::out | std::ios::binary);
    return trace.SerializeToOstream(&pb_ofile);
}

heracles::fhe_trace::Trace load_trace(const std::string &filename)
{
    std::ifstream pb_ifile(filename, std::ios::binary);
    heracles::fhe_trace::Trace trace;
    if (!trace.ParseFromIstream(&pb_ifile))
        throw std::runtime_error("Cannot read from file : " + filename);
    return trace;
}
} // namespace heracles::fhe_trace
